filingcabinet/views.py

```diff
@@ -178,6 +180,50 @@ def get_context_data(self, **kwargs):
         return context
 
 
+class DocumentCollectionZipDownloadView(AuthMixin, PkSlugMixin, DetailView):
+    model = DocumentCollection
+    redirect_url_name = "filingcabinet:document-collection_zip"
+    redirect_short_url_name = "filingcabinet:document-collection_zip_short"
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+
+        context.update(get_document_collection_context(self.object, self.request))
+        return context
+
+    def render_to_response(self, context):
+        archive_stream = zipstream.ZipFile(mode="w")
+        coll_docs = CollectionDocument.objects.filter(collection=self.object)
+        directory_dirname_map = {}
+        for doc in coll_docs:
+            if doc.directory is not None:
+                if doc.directory.pk not in directory_dirname_map:
+                    path_to_root = [
+                        *(x.name for x in doc.directory.get_ancestors()),
+                        doc.directory.name,
+                    ]
+                    dirname = "/".join(path_to_root)
+                    directory_dirname_map[doc.directory.pk] = dirname
+                else:
+                    dirname = directory_dirname_map[doc.directory.pk]
+                filename = "/".join(
+                    [
+                        dirname,
+                        doc.document.get_document_filename(),
+                    ]
+                )
+            else:
+                filename = doc.document.get_document_filename()
```

Comment:
Unfortunately, we do not guarantee that filenames in collection root or in directories are unique. This could be when a collection is assembled from similar document types. The filename is mostly a weird artifact of processes. The zipfile probably doesn't like multiple files with the same path?